master/internal/restore.go

```diff
@@ -0,0 +1,180 @@
+package internal
+
+import (
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+
+	"github.com/determined-ai/determined/master/internal/db"
+	"github.com/determined-ai/determined/master/internal/telemetry"
+	"github.com/determined-ai/determined/master/pkg/actor"
+	"github.com/determined-ai/determined/master/pkg/model"
+	"github.com/determined-ai/determined/master/pkg/searcher"
+)
+
+// The current experiment snapshot version. Once this is incremented, older versions should be
+// shimmed.
+const experimentSnapshotVersion = 0
+
+// Restore works by restoring from distributed consistent snapshots taken through the course
+// of an experiment. Snapshots within the system flow from the bottom up, starting with the
+// trial workload sequencer, to the trial, and finally to the experiment. Any event that the
+// trial or trial workload sequencer processes that would trigger a change to the state of the
+// experiment is:
+//   1. Propagated atomically, within a single message, to ensure the experiment handles it all
+//      or nothing
+//   2. With a snapshot affixed to it, to mark that it should trigger a snapshot
+// Upon receipt, the experiment handles the event entirely, snapshots its state and saves it
+// along with all the snapshots it has received, atomically.
+//
+// This is technically equivalent to the Chandy-Lamport distributed snapshotting algorithm where
+// the trial workload sequencer initiates the snapshots and propagates them along with trialCreated
+// and trialCompletedWorkload messages, with the snapshot itself acting as the marker. Each actor we
+// care about and know has updated state in the system snapshots it state when it receives this
+// marker and propagates it through the system. Even though all trials are part of each distributed
+// snapshot, we do not propagate snapshots to trials that didn't initiate the snapshot; that a trial
+// sends a snapshot with each state change implies if it did not send the snapshot, it does not have
+// a state change [1]. We can also be sure there are no pre-snapshot messages floating in the ether,
+// since any message we care about snapshotting is itself a snapshot.
+//
+// To restore properly, any message the experiment would send as a result of snapshot-able event
```

Comment:
> a searcher might send a couple of snapshot-able events to a trial but the system fails during handling one message.

within the current system, this is strictly not allowed. state updates can only flow in one direction now (other wise for the reasons you mentioned taking distributed snapshots gets much harder, because there can be messages you need in flight).

Reply:
in general with this scheme, all the snapshots are saved by the experiment once it gets them. at that point, we can be sure there are no important message left pertaining to message that caused the snapshot (since the event is all acknowledged in a single message).

Reply:
I see. It would be nice if you can put this constraint in the comment because it looks like a big constraint that we don't really have tests to enforce.

Reply:
yeah that's what I mean. it's in [1] in the bottom of the comment already, if that's what you were looking for.